src/boot/efi/xbootldr.c

```diff
@@ -0,0 +1,211 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include <efi.h>
+#include <efigpt.h>
+#include <efilib.h>
+
+#include "xbootldr.h"
+#include "util.h"
+
+static EFI_DEVICE_PATH *path_parent(EFI_DEVICE_PATH *path, EFI_DEVICE_PATH *node) {
+        EFI_DEVICE_PATH *parent;
+        UINTN len;
+
+        assert(path);
+        assert(node);
+
+        len = (UINT8*) NextDevicePathNode(node) - (UINT8*) path;
+        parent = (EFI_DEVICE_PATH*) AllocatePool(len + sizeof(EFI_DEVICE_PATH));
+        CopyMem(parent, path, len);
+        CopyMem((UINT8*) parent + len, EndDevicePath, sizeof(EFI_DEVICE_PATH));
+
+        return parent;
+}
+
+VOID xbootldr_open(EFI_HANDLE *device, EFI_HANDLE *ret_device, EFI_FILE **ret_root_dir) {
```

Comment:
splitting this out makes a ton of sense, but please, let's propagate errors here if we have any. i.e. the caller should ignore errors, not the callee.

Hence, change this to return EFI_STATUS please, and propagate the errors we see, or generate roughly reasonably errors for the cases we ourselves detect. i.e. if there's no xbootldr partition, then return `EFI_NOT_FOUND` and so on.

it's a matter of coding style i guess. across the codebase we do it that way: have inner functions propagate errors, and if they shall be eaten up then do so in outer functions. That helps with ensuring code remains as generic and reusable as we can.